components/locale/macros/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+
+fn get_value_from_token_stream(input: TokenStream) -> String {
+    let val = format!("{}", input);
+    if !val.starts_with('"') || !val.ends_with('"') {
+        panic!("Argument must be a string literal.");
+    }
+    let len = val.len();
+    (&val[1..len - 1]).to_string()
+}
+
+fn language_output(input: Option<u64>) -> String {
+    let val = if let Some(raw) = input {
+        format!("Some({})", raw)
+    } else {
+        "None".to_string()
+    };
+    format!("unsafe {{ icu_locale::subtags::Language::from_raw_unchecked({}) }}", val)
+}
+
+fn script_output(input: u32) -> String {
+    format!("unsafe {{ icu_locale::subtags::Script::from_raw_unchecked({}) }}", input)
+}
+
+fn region_output(input: u32) -> String {
+    format!("unsafe {{ icu_locale::subtags::Region::from_raw_unchecked({}) }}", input)
+}
+
+fn variant_output(input: u64) -> String {
+    format!("unsafe {{ icu_locale::subtags::Variant::from_raw_unchecked({}) }}", input)
+}
+
+#[proc_macro]
+pub fn language(input: TokenStream) -> TokenStream {
+    let val = get_value_from_token_stream(input);
+
+    let parsed: icu_locale::subtags::Language = val.parse().expect("Malformed Language Subtag");
+
+    let lang = language_output(parsed.into_raw());
+
+    lang.parse().unwrap()
+}
+
+#[proc_macro]
+pub fn script(input: TokenStream) -> TokenStream {
```

Comment:
These macros are public and should have rustdoc, right?

Reply:
Yes! Added!